File: backend/src/handlers/products/get_product.py
"""
Get Product Lambda Handler
GET /products/{id} - Get a single product by ID
"""
import json
import os
from typing import Any, Dict
from aws_lambda_powertools import Logger, Tracer
from aws_lambda_powertools.event_handler import APIGatewayHttpResolver
from aws_lambda_powertools.utilities.typing import LambdaContext
import boto3

logger = Logger()
tracer = Tracer()
app = APIGatewayHttpResolver()

dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table(os.environ['PRODUCTS_TABLE'])


@tracer.capture_method
def get_product_by_id(product_id: str) -> Dict[str, Any]:
    """Get a single product from DynamoDB."""
    
    try:
        response = table.get_item(
            Key={
                'PK': f'PRODUCT#{product_id}',
                'SK': 'METADATA'
            }
        )
        
        if 'Item' not in response:
            return None
        
        return response['Item']
        
    except Exception as e:
        logger.exception("Error fetching product")
        raise


@app.get("/products/<product_id>")
@tracer.capture_method
def get_product(product_id: str):
    """Handle GET /products/{id} request."""
    
    logger.info(f"Fetching product: {product_id}")
    
    try:
        product = get_product_by_id(product_id)
        
        if not product:
            return {
                'statusCode': 404,
                'body': json.dumps({
                    'error': 'NOT_FOUND',
                    'message': f'Product {product_id} not found'
                }),
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                }
            }
        
        # Check if product is active
        if product.get('status') != 'active':
            return {
                'statusCode': 404,
                'body': json.dumps({
                    'error': 'NOT_FOUND',
                    'message': 'Product not available'
                }),
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                }
            }
        
        return {
            'statusCode': 200,
            'body': json.dumps(product, default=str),
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Cache-Control': 'max-age=300'  # Cache for 5 minutes
            }
        }
        
    except Exception as e:
        logger.exception("Error processing request")
        return {
            'statusCode': 500,
            'body': json.dumps({
                'error': 'INTERNAL_ERROR',
                'message': 'An error occurred while processing your request'
            }),
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            }
        }


@logger.inject_lambda_context
@tracer.capture_lambda_handler
def handler(event: Dict[str, Any], context: LambdaContext) -> Dict[str, Any]:
    """Lambda handler entry point."""
    logger.info("Received event", extra={'event': event})
    return app.resolve(event, context)

File: backend/src/handlers/products/getProduct.js
const { DynamoDBClient } = require('@aws-sdk/client-dynamodb');
const { DynamoDBDocumentClient, GetCommand } = require('@aws-sdk/lib-dynamodb');

// Initialize DynamoDB client
const client = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(client);

// Environment variable
const PRODUCTS_TABLE = process.env.PRODUCTS_TABLE;

// CORS headers
const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type, Authorization'
};

/**
 * Lambda handler to get a single product by ID
 * @param {Object} event - API Gateway Lambda Proxy Input Format
 * @returns {Object} - API Gateway Lambda Proxy Output Format
 */
exports.handler = async (event) => {
    console.log('Event:', JSON.stringify(event, null, 2));

    try {
        // Validate environment variable
        if (!PRODUCTS_TABLE) {
            throw new Error('PRODUCTS_TABLE environment variable is not set');
        }

        // Extract product ID from path parameters
        const productId = event.pathParameters?.id;

        if (!productId) {
            return {
                statusCode: 400,
                headers,
                body: JSON.stringify({
                    message: 'Product ID is required'
                })
            };
        }

        // Construct DynamoDB key
        const key = {
            PK: `PRODUCT#${productId}`,
            SK: `PRODUCT#${productId}`
        };

        console.log('Fetching product with key:', JSON.stringify(key, null, 2));

        // Prepare GetItem parameters
        const params = {
            TableName: PRODUCTS_TABLE,
            Key: key
        };

        // Execute get operation
        const command = new GetCommand(params);
        const result = await ddbDocClient.send(command);

        // Check if product exists
        if (!result.Item) {
            return {
                statusCode: 404,
                headers,
                body: JSON.stringify({
                    message: `Product with ID '${productId}' not found`
                })
            };
        }

        console.log(`Successfully retrieved product: ${productId}`);

        return {
            statusCode: 200,
            headers,
            body: JSON.stringify(result.Item)
        };

    } catch (error) {
        console.error('Error retrieving product:', error);

        // Determine if it's a validation error or server error
        const statusCode = error.message.includes('environment variable') || 
                          error.message.includes('required') 
            ? 400 
            : 500;

        return {
            statusCode,
            headers,
            body: JSON.stringify({
                message: 'Failed to retrieve product',
                error: error.message
            })
        };
    }
};
